File: lab2/ConsoleApp1/T4.cs
namespace ConsoleApp1;

public class T4
{
    public static void Run(Data data)
    {
        Console.WriteLine("T4 is started");

        // Input d
        data.d = 1;

        // Input Z
        data.Z = Enumerable.Repeat(1, data.N).ToArray();

        // Input MM
        data.MM = new int[data.N, data.N];
        for (var i = 0; i < data.N; i++)
        for (var j = 0; j < data.N; j++)
            data.MM[i, j] = 1;

        // Signal T1, T2, T3 about input d, Z, MM 
        data.Sem2.Release(3);

        // Wait for input B, MX in T2
        data.Sem1.WaitOne();

        data.Mutex1.WaitOne();
        // Copy d4 = d
        var d4 = data.d;
        data.Mutex1.ReleaseMutex();

        // Calculation1 Ah = sort(d * Bh + Z * (MM * MXh))
        data.calculation1(d4, data.H * 3, data.H * 4);

        // Signal T3 about end of calculation Ah
        data.Sem5.Release();

        // Calculation4 ai = min(Bh)
        var a4 = data.calculation4(data.H * 3, data.H * 4);

        // Calculation5 a = min(a, ai)
        lock(data.aLock)
        {
            data.a = Math.Min(data.a, a4);
        }

        // Signal T1, T2, T3 about end of calculation a
        data.Event4.Set();

        // Wait for calculation a, A in T1
        data.Event1.WaitOne();

        // Wait for calculation a in T2
        data.Event2.WaitOne();

        // Wait for calculation a in T3
        data.Event3.WaitOne();

        data.Mutex2.WaitOne();
        // Copy a4 = a
        a4 = data.a;
        data.Mutex2.ReleaseMutex();

        // Calculation6 Xh = Ah * a
        data.calculation6(a4, data.H * 3, data.H * 4);

        // Wait for calculation Xh in T1, T2, T3
        data.Barrier.SignalAndWait();

        // Output X
        Console.WriteLine(string.Join(", ", data.X));

        Console.WriteLine("T4 is finished");
    }
}

File: lab1/src/T4.java
import java.util.Arrays;

public class T4 extends Thread {
    private int x4 = 0;

    public void run() {
        try {
            System.out.println("T4 is started");

            //Input R
            Arrays.fill(Data.R, 1);

            //Input Z
            Arrays.fill(Data.Z, 1);

            //Signal T1, T2, T3 about input MC, E
            Data.Sem4.release(3);

            //Wait for input MC, E in T1
            Data.Sem1.acquire(1);

            //Wait for input MD, d in T2
            Data.Sem2.acquire(1);

            //Wait for input B, p in T3
            Data.Sem3.acquire(1);

            //Calculation1 Sh = R * MCh
            Data.calculation1(Data.H * 3, Data.H * 4);

            //Calculation2 x4 = BH * ZH
            for (int i = 0; i < Data.H; i++) {
                x4 += Data.B[i] * Data.Z[i];
            }

            //Calculation3 x = x + x4
            Data.x.addAndGet(x4);

            //Signal T1, T2, T3 about end of calculation x
            Data.Sem8.release(3);

            //Wait for end of calculation x in T1
            Data.Sem5.acquire(1);

            //Wait for end of calculation x in T2
            Data.Sem6.acquire(1);

            //Wait for end of calculation x in T3
            Data.Sem7.acquire(1);

            //Copy p4 = p
            int p4;
            synchronized (Data.Y) {
                p4 = Data.p;
            }

            //Copy x4 = x
            Data.Sem0.acquire();
            x4 = Data.x.get();
            Data.Sem0.release();

            //Copy d4 = d
            int d4;
            synchronized (Data.W) {
                d4 = Data.d;
            }

            //Calculation4 Ah = S * MDh * p4 + x4 * Eh * d4
            Data.calculation4(p4, x4, d4, 0, Data.H);

            //Signal T3 about end of calculation Ah
            Data.Sem11.release();

            System.out.println("T4 is finished");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
